hudi-hadoop-mr/src/main/java/org/apache/hudi/hadoop/hive/HoodieCombineHiveInputFormat.java

```diff
@@ -375,6 +388,38 @@ public InputSplit[] getSplits(JobConf job, int numSplits) throws IOException {
     // clear work from ThreadLocal after splits generated in case of thread is reused in pool.
     Utilities.clearWorkMapForConf(job);
 
+    // build internal schema for the query
+    if (result.size() > 0) {
+      ArrayList<String> uniqTablePaths = new ArrayList<>();
+      Arrays.stream(paths).forEach(path -> {
+        HoodieStorage storage = null;
+        try {
+          storage = new HoodieHadoopStorage(path.getFileSystem(job));
+          Option<StoragePath> tablePath = TablePathUtils.getTablePath(storage, HadoopFSUtils.convertToStoragePath(path));
+          if (tablePath.isPresent()) {
+            uniqTablePaths.add(tablePath.get().toUri().toString());
+          }
+        } catch (IOException e) {
+          throw new RuntimeException(e);
+        }
+      });
+
+      try {
+        for (String path : uniqTablePaths) {
+          HoodieTableMetaClient metaClient = HoodieTableMetaClient.builder().setBasePath(path).setConf(new HadoopStorageConfiguration(job)).build();
+          TableSchemaResolver schemaUtil = new TableSchemaResolver(metaClient);
+          Option<InternalSchema> schema = schemaUtil.getTableInternalSchemaFromCommitMetadata();
+          if (schema.isPresent()) {
+            LOG.info("Set internal schema and avro schema of path: " + path.toString());
+            job.set(INTERNAL_SCHEMA_CACHE_KEY_PREFIX + "." + path, SerDeHelper.toJson(schema.get()));
+            job.set(SCHEMA_CACHE_KEY_PREFIX + "." + path, schemaUtil.getTableAvroSchema().toString());
```

Comment:
We cache the schema by path or by table name? The table name looks more straight-forward. And the schemaUtil.getTableAvroSchema() should be invoked first to set up the commit metadata cache to avoid redundant commit metadata deserialization.

Reply:
@danny0405 That's a great suggestion. When we handle tables from two different databases together, there might be the same table name. For example, dev.table_xx and online.table_xx, where table_xx is the same. Should we consider this case?

Reply:
I think we can use table name now, can extent it to path if needed.